In [16]:
import pandas as pd
import numpy as np
import re

In [17]:
patients=pd.read_csv('patients.csv')
treatment=pd.read_csv('treatments.csv')
adverse=pd.read_csv('adverse_reactions.csv')

In [18]:
print(patients.shape)
print(treatment.shape)
print(adverse.shape)

(503, 14)
(280, 7)
(34, 3)


In [19]:
patients

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
..                         ...                ...         ...       ...   
498        2530 Victoria Court       Milton Mills          ME    3852.0   
499   494 Clarksburg Park Road             Sedona          AZ   86341.0   
500          649 Nutter Street      Overland Park          MO   64110.0   
501   3652 Boone Crockett Lane            Seattle          WA   98109.0   
502          2778 North Avenue               Burr    Nebraska   68324.0   

           country                                          contact  \
0    United States             951-719-9170ZoeWellish@superrito.com   
1    United States            PamelaSHill@cuvox.de+1 (217) 569-3204   
2    United States                 402-363-6804JaeMDebord@gustr.com   
3    United States       PhanBaLiem@jourrapide.com+1 (732) 636-8246   
4    United States                  334-515-7487TimNeudorf@cuvox.de   
..             ...                                              ...   
498  United States      207-477-0579MustafaLindstrom@jourrapide.com   
499  United States               928-284-4492RumanBisliev@gustr.com   
500  United States            816-223-6007JinkedeKeizer@teleworm.us   
501  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060   
502  United States             PatrickGersten@rhyta.com402-848-4923   

     birthdate  weight  height   bmi  
0    7/10/1976   121.7      66  19.6  
1     4/3/1967   118.8      66  19.2  
2    2/19/1980   177.8      71  24.8  
3    7/26/1951   220.9      70  31.7  
4    2/18/1928   192.3      27  26.1  
..         ...     ...     ...   ...  
498  4/10/1959   181.1      72  24.6  
499  3/26/1948   239.6      70  34.4  
500  1/13/1971   171.2      67  26.8  
501  2/13/1952   176.9      67  27.7  
502   5/3/1954   138.2      71  19.3  

[503 rows x 14 columns]

In [20]:
treatment

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika      jindrová  41u - 48u          -         7.63       7.20   
1       elliot    richardson          -  40u - 45u         7.56       7.09   
2     yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3         skye    gormanston  33u - 36u          -         7.97       7.62   
4       alissa        montez          -  33u - 29u         7.78       7.46   
..         ...           ...        ...        ...          ...        ...   
275     albina      zetticci  45u - 51u          -         7.93       7.73   
276       john   teichelmann          -  49u - 49u         7.90       7.58   
277     mathea       lillebø  23u - 36u          -         9.04       8.67   
278     vallie        prince  31u - 38u          -         7.64       7.28   
279     samúel  guðbrandsson  53u - 56u          -         8.00       7.64   

     hba1c_change  
0             NaN  
1            0.97  
2             NaN  
3            0.35  
4            0.32  
..            ...  
275          0.20  
276           NaN  
277          0.37  
278          0.36  
279          0.36  

[280 rows x 7 columns]

In [21]:
adverse.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

In [22]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [23]:
treatment.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [24]:
adverse.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
given_name          34 non-null object
surname             34 non-null object
adverse_reaction    34 non-null object
dtypes: object(3)
memory usage: 944.0+ bytes


In [25]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [26]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [27]:
treatment.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [28]:
patients.sample(5)

patient_id assigned_sex given_name      surname               address  \
270         271       female     Jowita   Wiśniewska   2168 Butternut Lane   
161         162         male  Mitsunobu     Fujiwara      212 Tibbs Avenue   
380         381       female      Silje  Kristiansen  4992 Moonlight Drive   
27           28         male     Joseph       Tucker      4982 Wood Street   
325         326       female       Sara  Karjalainen  720 Tator Patch Road   

              city     state  zip_code        country  \
270   Granite City  Illinois   62040.0  United States   
161        Chester        MT   59522.0  United States   
380  Pleasantville        NJ    8232.0  United States   
27          Venice        LA   70091.0  United States   
325        Chicago  Illinois   60601.0  United States   

                                           contact  birthdate  weight  height  \
270  JowitaWisniewska@armyspy.com+1 (618) 512-3319  11/8/1934   108.1      61   
161      406-759-6160MitsunobuFujiwara@armyspy.com  5/21/1969   223.5      67   
380       SiljeAKristiansen@dayrep.com609-641-2070   2/7/1996   162.4      64   
27             985-814-7603JosephNTucker@rhyta.com  4/10/1959   175.8      72   
325      SaraKarjalainen@armyspy.com1 312 856 6826  8/20/1946   172.5      63   

      bmi  
270  20.4  
161  35.0  
380  27.9  
27   23.8  
325  30.6

In [29]:
patients[patients['address'].duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

### Quality

[Completeness, Validity, Accuracy, Consistency]

#### `Patients`

- Zip code col is stored in decimal format[V]
- Certain zip codes of 4 digits[V] 
- For the state col sometimes full names are use and other times short forms are used[IC]
- Missing values for contact details of 12 patients[C]
- Incorrect data types for assigned_sex and birthdate cols[?]
- Erroneous data for weights and heights[Accuracy]
- Inconsistent way of storing phone numbers in contact col[IC]
- Duplicate and not useful data in dataset[IC]

#### `Treatments`

- hba1c_change col has missing values[C]
- Novodra and Auralin cols has u attached to numbers[V]
- Only 280 rows are there, instead of 350[C]
- Name and surname col is in lower case[IC]
- Incorrect data type for auralin and novadra cols[V]
- hba1c_change 9 instead of 4[Accuracy]
- use of dashes instead of Nan in aurolin and novodra cols[IC]

#### `Adverse Reactions`

- Name and surname col is in lower case[IC]

### Tidiness

#### `Patients`

- Phone and email together in one col i.e. contact

#### `Treatments`

- 3 variables stored in auralin and novodra cols

#### `Adverse Reactions`

- No point of using this table

In [31]:
patients_copy[patients_copy['given_name']=='John']

patient_id assigned_sex given_name      surname                  address  \
140         141         male       John  Teichelmann    631 Isaacs Creek Road   
153         154         male       John     Carreiro  1463 Martha Ellen Drive   
215         216         male       John          Doe          123 Main Street   
229         230         male       John          Doe          123 Main Street   
237         238         male       John          Doe          123 Main Street   
242         243         male       John      O'Brian                      NaN   
244         245         male       John          Doe          123 Main Street   
251         252         male       John          Doe          123 Main Street   
277         278         male       John          Doe          123 Main Street   

         city state  zip_code        country  \
140   Decatur    IL   62522.0  United States   
153      Reno    NV   89509.0  United States   
215  New York    NY   12345.0  United States   
229  New York    NY   12345.0  United States   
237  New York    NY   12345.0  United States   
242       NaN   NaN       NaN            NaN   
244  New York    NY   12345.0  United States   
251  New York    NY   12345.0  United States   
277  New York    NY   12345.0  United States   

                                       contact  birthdate  weight  height  \
140     217-421-0935JohnTeichelmann@dayrep.com  12/3/1958   214.1      72   
153  JohnACarreiro@superrito.com1 775 770 7827   4/8/1976   177.8      68   
215                johndoe@email.com1234567890   1/1/1975   180.0      72   
229                johndoe@email.com1234567890   1/1/1975   180.0      72   
237                johndoe@email.com1234567890   1/1/1975   180.0      72   
242                                        NaN  2/25/1957   205.3      74   
244                johndoe@email.com1234567890   1/1/1975   180.0      72   
251                johndoe@email.com1234567890   1/1/1975   180.0      72   
277                johndoe@email.com1234567890   1/1/1975   180.0      72   

      bmi  
140  29.0  
153  27.0  
215  24.4  
229  24.4  
237  24.4  
242  26.4  
244  24.4  
251  24.4  
277  24.4

In [58]:

patients_copy[patients_copy.duplicated(subset=['given_name','surname','Emails','Phone numbers','address','assigned_sex','bmi','city','country','height','state'])]

Emails Phone numbers          address assigned_sex birthdate  \
227  johndoe@email.com    1234567890  123 Main Street         male  1/1/1975   
233  johndoe@email.com    1234567890  123 Main Street         male  1/1/1975   
239  johndoe@email.com    1234567890  123 Main Street         male  1/1/1975   
245  johndoe@email.com    1234567890  123 Main Street         male  1/1/1975   
268  johndoe@email.com    1234567890  123 Main Street         male  1/1/1975   

      bmi      city        country given_name  height  patient_id state  \
227  24.4  New York  United States       John      72         230    NY   
233  24.4  New York  United States       John      72         238    NY   
239  24.4  New York  United States       John      72         245    NY   
245  24.4  New York  United States       John      72         252    NY   
268  24.4  New York  United States       John      72         278    NY   

    surname  weight  zip_code  
227     Doe   180.0   12345.0  
233     Doe   180.0   12345.0  
239     Doe   180.0   12345.0  
245     Doe   180.0   12345.0  
268     Doe   180.0   12345.0

In [59]:
patients_copy[patients_copy['Emails']!='johndoe@email.com']

Emails      Phone numbers                    address  \
0     ZoeWellish@superrito.com       951-719-9170       576 Brown Bear Drive   
1         PamelaSHill@cuvox.de  +1 (217) 569-3204  2370 University Hill Road   
2         JaeMDebord@gustr.com       402-363-6804      1493 Poling Farm Road   
3    PhanBaLiem@jourrapide.com  +1 (732) 636-8246        2335 Webster Street   
4          TimNeudorf@cuvox.de       334-515-7487       1428 Turkey Pen Lane   
..                         ...                ...                        ...   
498                        NaN                NaN                        NaN   
499                        NaN                NaN                        NaN   
500                        NaN                NaN                        NaN   
501                        NaN                NaN                        NaN   
502                        NaN                NaN                        NaN   

    assigned_sex   birthdate   bmi               city        country  \
0         female   7/10/1976  19.6  Rancho California  United States   
1         female    4/3/1967  19.2          Armstrong  United States   
2           male   2/19/1980  24.8               York  United States   
3           male   7/26/1951  31.7         Woodbridge  United States   
4           male   2/18/1928  26.1             Dothan  United States   
..           ...         ...   ...                ...            ...   
498       female   11/3/1989  28.1                NaN            NaN   
499       female   10/9/1937  33.1                NaN            NaN   
500       female  12/16/1962  18.4                NaN            NaN   
501         male    4/1/1979  23.6                NaN            NaN   
502       female   5/14/1990  32.1                NaN            NaN   

    given_name  height  patient_id       state     surname  weight  zip_code  
0          Zoe      66           1  California     Wellish   121.7   92390.0  
1       Pamela      66           2    Illinois        Hill   118.8   61812.0  
2          Jae      71           3    Nebraska      Debord   177.8   68467.0  
3         Liêm      70           4          NJ        Phan   220.9    7095.0  
4          Tim      27           5          AL     Neudorf   192.3   36303.0  
..         ...     ...         ...         ...         ...     ...       ...  
498   Wafiyyah      63         265         NaN      Asfour   158.6       NaN  
499     Flavia      61         270         NaN  Fiorentino   175.2       NaN  
500   Generosa      69         279         NaN       Cabán   124.3       NaN  
501      Lewis      68         287         NaN        Webb   155.3       NaN  
502        Chỉ      63         297         NaN         Lâm   181.1       NaN  

[497 rows x 15 columns]


# Cleaning Data

You should follow the order provided below:

- Completeness Issues
- Tidiness Issues
- Remaining Data quality issues like validity, accuracy and consistency

### Steps involved in Data cleaning

- Define
- Code 
- Test

`Always make sure to create a copy of your pandas dataframe before you start the cleaning process`

In [111]:
# Creating copies

patients_copy=patients.copy()
treatment_copy=treatment.copy()
adverse_copy=adverse.copy()

#Importing the rest of treatment data
treatment_rest=pd.read_csv('treatments_cut.csv')

In [112]:
patients_copy[patients_copy['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

## a. Completeness Issues

- Missing values for contact details of 12 patients
- Only 280 rows are there, instead of 350
- hba1c_change col has missing values

#### Define

In [43]:

#Concatinate the treatement dataset to give 350 values


#### Code

In [113]:
treatment_copy=pd.concat([treatment_copy,treatment_rest])

#### Test

In [114]:

treatment_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 69
Data columns (total 7 columns):
given_name      350 non-null object
surname         350 non-null object
auralin         350 non-null object
novodra         350 non-null object
hba1c_start     350 non-null float64
hba1c_end       350 non-null float64
hba1c_change    213 non-null float64
dtypes: float64(3), object(4)
memory usage: 21.9+ KB


#### Problem

In [115]:
treatment_copy[treatment_copy.duplicated()] 

given_name surname    auralin novodra  hba1c_start  hba1c_end  \
136     joseph     day  29u - 36u       -          7.7       7.19   

     hba1c_change  
136           NaN

## b. Tidiness Issues

- Phone + email in contact col `patients`
- Novodra and Aurolin col `treatment`
- Merge the `adverse` table to `treatment` table

#### Define

In [116]:
#contact and email in same column in patients


#### Code

In [117]:
unknown=patients_copy[patients_copy['address'].isnull()]
patients_copy.dropna(inplace=True)

In [118]:
emails=[]
numbers=[]
for i in patients_copy['contact']:
    pattern=re.compile(r'[a-zA-Z][a-zA-Z0-9.-]+[@][a-zA-Z0-9-]+\.[a-zA-Z0-9.-]+[a-zA-Z]')
    numbers.append(i.replace(pattern.findall(i)[0],''))
    emails.append(pattern.findall(i)[0])

In [119]:
patients_copy['Emails']=emails
patients_copy['Phone numbers']=numbers

In [120]:
patients_copy=pd.concat([patients_copy,unknown]).reset_index()

C:\Users\Sanaa\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """Entry point for launching an IPython kernel.


In [121]:
patients_copy=patients_copy.drop(columns=['index','contact'])

#### Test

In [122]:
patients_copy.sample(10)

Emails      Phone numbers  \
73                ZarkaRap@superrito.com       978-969-6753   
391             RobertMaslov@fleckens.hu       401-535-2675   
143  AmandaCavalcantiRibeiro@fleckens.hu       206-902-0473   
210          MarthaAfanasyeva@dayrep.com     1 612 228 4170   
397        AlbikaIbragimov@superrito.com       716-897-5050   
387                EricEk@jourrapide.com  +1 (360) 965-1558   
112    ArnruurSkarpheinsdottir@rhyta.com  +1 (757) 885-5669   
288            BenignaFerrari@einrot.com       706-616-0152   
236           MarphisaCompagnon@cuvox.de  +1 (256) 615-5522   
119                 ChajaBouw@dayrep.com       832-457-6625   

                    address assigned_sex   birthdate   bmi          city  \
73      2661 Russell Street       female  12/28/1995  21.2       Foxboro   
391        2356 Myra Street         male    7/2/1932  36.5    Providence   
143        4155 Raccoon Run       female   9/27/1970  25.6       Seattle   
210    3613 Lodgeville Road       female   10/3/1997  19.9    Saint Paul   
397      4945 Jarvis Street       female   2/29/1948  28.6       Buffalo   
387       4160 Pratt Avenue         male    4/7/1943  29.6       Seattle   
112            2578 Tenmile       female    6/4/1922  30.5  Newport News   
288  4912 Hart Country Lane       female    1/6/1960  34.3      Columbus   
236      3391 Marcus Street       female    3/9/1967  35.1    Huntsville   
119       2808 Circle Drive       female   9/16/1979  30.4       Houston   

           country given_name  height  patient_id     state  \
73   United States      Žarka      68          74        MA   
391  United States     Robert      65         404        RI   
143  United States     Amanda      60         144        WA   
210  United States     Martha      73         212        MN   
397  United States     Albika      63         410  New York   
387  United States       Eric      71         400        WA   
112  United States  Arnþrúður      63         113        VA   
288  United States    Benigna      61         301        GA   
236  United States   Marphisa      63         241        AL   
119  United States      Chaja      70         120        TX   

               surname  weight  zip_code  
73                 Rap   139.3    2035.0  
391             Maslov   219.1    2908.0  
143            Ribeiro   131.1   98109.0  
210         Afanasyeva   151.2   55114.0  
397          Ibragimov   161.7   14211.0  
387                 Ek   212.1   98109.0  
112  Skarphéðinsdóttir   172.0   23602.0  
288            Ferrari   181.5   31901.0  
236          Compagnon   198.0   35806.0  
119               Bouw   211.9   77063.0

#### Define

In [123]:
# Novodra and Aurolin using the melt function

#### Code

In [124]:
treatment_copy=pd.melt(treatment_copy, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
                           var_name='treatment', value_name='dose')

In [125]:
treatment_copy=treatment_copy[treatment_copy['dose']!='-']

In [126]:
treatment_copy['start'],treatment_copy['end']=treatment_copy['dose'].str.split('-').str

In [127]:
treatment_copy.drop(columns={'dose'}, inplace=True)

#### Test

In [128]:
treatment_copy.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20           NaN   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37           NaN   auralin   

  start   end  
0  41u    48u  
3  33u    36u  
6  37u    42u  
7  31u    38u  
9  30u    36u

#### Define

In [129]:
# Adverse Dataset not useful(so merging it with treatment)

#### Code

In [130]:
treatment_copy=pd.merge(left=treatment_copy, right=adverse_copy, how='left', on=['given_name','surname'])

#### Test

In [131]:
treatment_copy.sample(10)


given_name      surname  hba1c_start  hba1c_end  hba1c_change treatment  \
138  inunnguaq     heilmann         7.85       7.45           NaN   auralin   
247      vasco  van de wiel         7.94       7.51          0.93   novodra   
217     hideki    haraguchi         7.59       7.05          0.54   novodra   
101    maximus       henzen         7.79       7.34           NaN   auralin   
298    sargent      flamand         7.92       7.49          0.93   novodra   
214  mackenzie        mckay         9.87       9.48          0.39   novodra   
279        emy       markus         7.53       7.14          0.39   novodra   
285       diệt          bùi         9.11       8.76          0.35   novodra   
81      lamara     dratchev         7.75       7.40          0.35   auralin   
202    albinca      komavec         7.89       7.46           NaN   novodra   

    start   end adverse_reaction  
138  57u    67u              NaN  
247  31u    29u              NaN  
217  37u    35u              NaN  
101  49u    55u              NaN  
298  47u    39u              NaN  
214  44u    43u              NaN  
279  29u    24u              NaN  
285  30u    32u              NaN  
81   49u    58u              NaN  
202  41u    39u     hypoglycemia

## Validity and Accuracy issues
- Novodra and Auralin cols has u attached to numbers[V]
- Incorrect data type for zip code [v]
- hba1c_change has incorrect values [A]
- some zip codes have 4 digits [V](#unable to correct this due to incorrect data)
- Incorrect data type for auralin and novadra cols[V] 


#### Define

In [132]:
#Novodra and Auralin cols has u attached to numbers

#### Code

In [133]:
start=[]
end=[]
for i in range(0,len(treatment_copy)):
    start.append(int(treatment_copy['start'][i].split('u')[0]))
    end.append(int(treatment_copy['end'][i].split('u')[0]))


In [134]:
treatment_copy['start']=start
treatment_copy['end']=end

#### Test

In [135]:
treatment_copy.sample(5)

given_name    surname  hba1c_start  hba1c_end  hba1c_change treatment  \
227       abel    yonatan         7.88       7.50           NaN   novodra   
121      tomáš   navrátil         7.84       7.41           NaN   auralin   
236      suoma   koivunen         7.59       7.13           NaN   novodra   
254    mustafa  lindström         7.92       7.60           NaN   novodra   
75     rebecca     priest         7.87       7.57           0.3   auralin   

     start  end adverse_reaction  
227     38   39            cough  
121     24   36              NaN  
236     26   25              NaN  
254     35   33              NaN  
75      40   42              NaN

In [136]:
treatment_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 349
Data columns (total 9 columns):
given_name          350 non-null object
surname             350 non-null object
hba1c_start         350 non-null float64
hba1c_end           350 non-null float64
hba1c_change        213 non-null float64
treatment           350 non-null object
start               350 non-null int64
end                 350 non-null int64
adverse_reaction    35 non-null object
dtypes: float64(3), int64(2), object(4)
memory usage: 37.3+ KB


#### Define

In [137]:
 # hba1c_change has incorrect values 

#### Code

In [138]:
treatment_copy['hba1c_change']=treatment_copy['hba1c_start']-treatment_copy['hba1c_end']

#### Test

In [139]:
treatment_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 349
Data columns (total 9 columns):
given_name          350 non-null object
surname             350 non-null object
hba1c_start         350 non-null float64
hba1c_end           350 non-null float64
hba1c_change        350 non-null float64
treatment           350 non-null object
start               350 non-null int64
end                 350 non-null int64
adverse_reaction    35 non-null object
dtypes: float64(3), int64(2), object(4)
memory usage: 37.3+ KB


In [140]:

treatment_copy.info()

treatment_copy.describe()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 349
Data columns (total 9 columns):
given_name          350 non-null object
surname             350 non-null object
hba1c_start         350 non-null float64
hba1c_end           350 non-null float64
hba1c_change        350 non-null float64
treatment           350 non-null object
start               350 non-null int64
end                 350 non-null int64
adverse_reaction    35 non-null object
dtypes: float64(3), int64(2), object(4)
memory usage: 37.3+ KB


hba1c_start   hba1c_end  hba1c_change       start         end
count   350.000000  350.000000    350.000000  350.000000  350.000000
mean      7.956343    7.560057      0.396286   39.071429   43.045714
std       0.545328    0.545456      0.059283    9.309885   10.523989
min       7.500000    7.010000      0.200000   19.000000   21.000000
25%       7.650000    7.270000      0.350000   32.000000   36.000000
50%       7.785000    7.400000      0.390000   39.000000   42.000000
75%       7.950000    7.557500      0.440000   45.000000   50.000000
max       9.950000    9.580000      0.540000   63.000000   75.000000

#### Define

In [141]:
# Zip code is in decimal(invalid)

#### Code

In [142]:

missing=patients_copy[patients_copy['zip_code'].isnull()]
patients_copy=patients_copy.dropna(subset=['zip_code']).astype({'zip_code':int})
patients_copy=patients_copy.dropna(subset=['zip_code']).astype({'zip_code':str})

In [143]:

patients_copy=pd.concat([patients_copy,missing]).reset_index()

#### Test

In [144]:
patients_copy.head()

index                     Emails      Phone numbers  \
0      0   ZoeWellish@superrito.com       951-719-9170   
1      1       PamelaSHill@cuvox.de  +1 (217) 569-3204   
2      2       JaeMDebord@gustr.com       402-363-6804   
3      3  PhanBaLiem@jourrapide.com  +1 (732) 636-8246   
4      4        TimNeudorf@cuvox.de       334-515-7487   

                     address assigned_sex  birthdate   bmi               city  \
0       576 Brown Bear Drive       female  7/10/1976  19.6  Rancho California   
1  2370 University Hill Road       female   4/3/1967  19.2          Armstrong   
2      1493 Poling Farm Road         male  2/19/1980  24.8               York   
3        2335 Webster Street         male  7/26/1951  31.7         Woodbridge   
4       1428 Turkey Pen Lane         male  2/18/1928  26.1             Dothan   

         country given_name  height  patient_id       state  surname  weight  \
0  United States        Zoe      66           1  California  Wellish   121.7   
1  United States     Pamela      66           2    Illinois     Hill   118.8   
2  United States        Jae      71           3    Nebraska   Debord   177.8   
3  United States       Liêm      70           4          NJ     Phan   220.9   
4  United States        Tim      27           5          AL  Neudorf   192.3   

  zip_code  
0    92390  
1    61812  
2    68467  
3     7095  
4    36303

## Inconsistency in Data
- For the state col sometimes full names are use and other times short forms are used[IC]
- Inconsistent way of storing phone numbers in contact col[IC]
- Duplicate and not useful data in dataset[IC]
- Name and surname col is in lower case[IC]
- use of dashes instead of Nan in aurolin and novodra cols[IC]



#### Define

In [145]:
# Inconsistensy in the names of states(can't do this)


#### Code

#### Test

#### Define

In [146]:

#given name and surname start with small letter

#### Code

In [147]:

first=[]
last=[]
for i in treatment_copy['given_name'].values:
    first.append(i.title())
for i in treatment_copy['surname'].values:
    last.append(i.title())
treatment_copy['given_name']=first
treatment_copy['surname']=last

#### Test

In [148]:

treatment_copy.sample(5)

given_name   surname  hba1c_start  hba1c_end  hba1c_change treatment  \
299         Wu      Sung         7.61       7.12          0.49   novodra   
44       Karen  Jakobsen         7.59       7.25          0.34   auralin   
224      Kwasi     Mensa         8.29       7.89          0.40   novodra   
176   Yukitaka  Takenaka         7.68       7.25          0.43   novodra   
133     Albina  Zetticci         7.93       7.73          0.20   auralin   

     start  end adverse_reaction  
299     47   48              NaN  
44      34   42              NaN  
224     40   36              NaN  
176     39   36              NaN  
133     45   51              NaN

#### Define

In [149]:
#presence of placeholders ('John Doe')
patients_copy[patients_copy.duplicated(subset=['given_name','surname','Emails','Phone numbers','address','assigned_sex','bmi','city','country','height','state'])]

index             Emails Phone numbers          address assigned_sex  \
227    227  johndoe@email.com    1234567890  123 Main Street         male   
233    233  johndoe@email.com    1234567890  123 Main Street         male   
239    239  johndoe@email.com    1234567890  123 Main Street         male   
245    245  johndoe@email.com    1234567890  123 Main Street         male   
268    268  johndoe@email.com    1234567890  123 Main Street         male   

    birthdate   bmi      city        country given_name  height  patient_id  \
227  1/1/1975  24.4  New York  United States       John      72         230   
233  1/1/1975  24.4  New York  United States       John      72         238   
239  1/1/1975  24.4  New York  United States       John      72         245   
245  1/1/1975  24.4  New York  United States       John      72         252   
268  1/1/1975  24.4  New York  United States       John      72         278   

    state surname  weight zip_code  
227    NY     Doe   180.0    12345  
233    NY     Doe   180.0    12345  
239    NY     Doe   180.0    12345  
245    NY     Doe   180.0    12345  
268    NY     Doe   180.0    12345

#### Code

In [150]:
patients_copy=patients_copy[patients_copy['Emails']!='johndoe@email.com']

#### Test 

In [151]:
patients_copy[patients_copy.duplicated(subset=['given_name','surname','Emails','Phone numbers','address','assigned_sex','bmi','city','country','height','state'])]

Empty DataFrame
Columns: [index, Emails, Phone numbers, address, assigned_sex, birthdate, bmi, city, country, given_name, height, patient_id, state, surname, weight, zip_code]
Index: []

In [152]:
patients_copy[patients_copy['Emails']=='johndoe@email.com']

Empty DataFrame
Columns: [index, Emails, Phone numbers, address, assigned_sex, birthdate, bmi, city, country, given_name, height, patient_id, state, surname, weight, zip_code]
Index: []

#### Final Dataset after Analysis into csv 

In [153]:

patients_copy.to_csv('patients_Final.csv')

In [154]:
treatment_copy.to_csv('treatment_final.csv')

### What is Assessing Data?

The data analysis process consists of the following steps

1. Asking Questions
2. Data Wrangling<br><br>
    a. Gathering Data
            i.   CSV files
            ii.  APIs
            iii. Web Scraping
            iv.  Databases
    b. <b>Assessing Data</b><br>
    c. Cleaning Data<br><br>
3. Exploratory Data Analysis
4. Drawing Conclusion
5. Comunicating Results
    

### Assessing Data

In this step, the data is to be understood more deeply. Before implementing methods to clean it, you will definitely need to have a better idea about what the data is about.

### Types of Unclean Data

There are 2 kinds of unclean data

1. <u>Dirty Data</u> (Data with Quality issues):<br>
    Dirty data, also known as low quality data. Low quality data has content issues.<br>
2. <u>Messy Data</u> (Data with tidiness issues)<br>
    Messy data, also known as untidy data. Untidy data has structural issues.

### 1. Dirty Data

When the data has following kinds of issues:

1. Duplicated data
2. Missing Data
3. Corrupt Data
4. Inaccurate Data

This kind of data is known as Dirty data

### 2. Messy Data

When the data has structural issue is it known as Messy or untidy data.

Tidy data has the following properties:

1. Each variable forms a column
2. Each observation forms a row
3. Each observational unit forms a table

If a data violates any one of the above it is known as untidy data

Example:

<img src="untidy.png">

### Types of Assessment

There are 2 types of assessment styles

1. Manual - Looking through the data manually in google sheets
2. Programmatic - By using pandas functions such as info(), describe() or sample()

### Steps in Assessment

There are 2 steps involved in Assessment

1. Discover
2. Document

### Data Quality Dimensions

1. Completeness Issues
2. Validity Issues
3. Accuracy Issues
4. Consistency Issues